File: generate_dreamcore.js
#!/usr/bin/env node
/**
 * Dreamcore LUT Generator for Nothing Camera
 * Translates exact Lightroom parameters into a standard 33x33x33 .CUBE 3D LUT
 */

const fs = require('fs');

function clamp(val, min = 0.0, max = 1.0) {
  return Math.max(min, Math.min(max, val));
}

// Convert RGB [0..1] to HSL [h in 0..360, s in 0..1, l in 0..1]
function rgbToHsl(r, g, b) {
  const max = Math.max(r, g, b);
  const min = Math.min(r, g, b);
  let h = 0;
  let s = 0;
  const l = (max + min) / 2;

  if (max !== min) {
    const d = max - min;
    s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
    switch (max) {
      case r: h = ((g - b) / d + (g < b ? 6 : 0)); break;
      case g: h = ((b - r) / d + 2); break;
      case b: h = ((r - g) / d + 4); break;
    }
    h *= 60;
  }
  return [h, s, l];
}

function hueToRgb(p, q, t) {
  if (t < 0) t += 1;
  if (t > 1) t -= 1;
  if (t < 1/6) return p + (q - p) * 6 * t;
  if (t < 1/2) return q;
  if (t < 2/3) return p + (q - p) * (2/3 - t) * 6;
  return p;
}

// Convert HSL [h in 0..360, s in 0..1, l in 0..1] to RGB [0..1]
function hslToRgb(h, s, l) {
  h = (h % 360 + 360) % 360;
  const hNorm = h / 360;
  if (s === 0) {
    return [l, l, l];
  }
  const q = l < 0.5 ? l * (1 + s) : l + s - l * s;
  const p = 2 * l - q;
  const r = hueToRgb(p, q, hNorm + 1/3);
  const g = hueToRgb(p, q, hNorm);
  const b = hueToRgb(p, q, hNorm - 1/3);
  return [r, g, b];
}

// 6 Control points for Tone Curve (Lifted blacks, matte look, smooth highlight ceiling)
const curveX = [0.00, 0.18, 0.38, 0.62, 0.82, 1.00];
const curveY = [0.08, 0.22, 0.42, 0.63, 0.83, 0.97];

function evaluateToneCurve(val) {
  val = clamp(val, 0.0, 1.0);
  for (let i = 0; i < curveX.length - 1; i++) {
    if (val >= curveX[i] && val <= curveX[i + 1]) {
      const t = (val - curveX[i]) / (curveX[i + 1] - curveX[i]);
      // Smooth Hermite interpolation between control points
      const smoothT = t * t * (3 - 2 * t);
      return curveY[i] + smoothT * (curveY[i + 1] - curveY[i]);
    }
  }
  return val;
}

// Hue center targets and falloff
const HUE_CENTERS = {
  red: 0.0,
  orange: 30.0,
  yellow: 60.0,
  green: 120.0,
  aqua: 180.0,
  blue: 240.0,
  purple: 285.0,
  magenta: 325.0
};

function getHueWeights(hue) {
  h = (hue % 360 + 360) % 360;
  let weights = {};
  let total = 0;
  for (const [name, center] of Object.entries(HUE_CENTERS)) {
    let diff = Math.abs(h - center);
    if (diff > 180) diff = 360 - diff;
    const w = Math.max(0, 1.0 - diff / 35.0);
    weights[name] = w;
    total += w;
  }
  if (total > 0) {
    for (const name in weights) {
      weights[name] /= total;
    }
  }
  return weights;
}

function processColor(inR, inG, inB) {
  let [r, g, b] = [inR, inG, inB];

  // 1. White Balance (Temp: -8, Tint: +5)
  const temp = -8.0;
  const tint = 5.0;
  const rGain = 1.0 + (temp / 100.0) * 0.12 + (tint / 100.0) * 0.08;
  const gGain = 1.0 - (tint / 100.0) * 0.08;
  const bGain = 1.0 - (temp / 100.0) * 0.12 - (tint / 100.0) * 0.04;
  r = clamp(r * rGain);
  g = clamp(g * gGain);
  b = clamp(b * bGain);

  // 2. Exposure (+0.60 EV)
  const exposure = 0.60;
  const expFactor = Math.pow(2.0, exposure);
  r *= expFactor;
  g *= expFactor;
  b *= expFactor;

  // 3. Dehaze (-22: adds airy haze by lifting floor and softening micro-contrast)
  const dehaze = -22.0;
  const haze = (dehaze / 100.0) * 0.12;
  r = clamp(r - haze * (1.0 - r));
  g = clamp(g - haze * (1.0 - g));
  b = clamp(b - haze * (1.0 - b));

  // 4. Basic Contrast (-30)
  const contrast = -30.0;
  const c = (contrast / 100.0) * 0.40;
  r = clamp((r - 0.45) * (1.0 + c) + 0.45);
  g = clamp((g - 0.45) * (1.0 + c) + 0.45);
  b = clamp((b - 0.45) * (1.0 + c) + 0.45);

  // 5. Highlights (-54), Shadows (+34), Whites (+13), Blacks (+78)
  const luma = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  const highlights = -54.0;
  const shadows = 34.0;
  const whites = 13.0;
  const blacks = 78.0;

  const hMask = clamp((luma - 0.45) / 0.55);
  const sMask = clamp((0.55 - luma) / 0.55);
  const wMask = clamp(luma);
  const bMask = clamp(1.0 - luma);

  const hAdj = (highlights / 100.0) * 0.35 * hMask;
  const sAdj = (shadows / 100.0) * 0.30 * sMask;
  const wAdj = (whites / 100.0) * 0.15 * (wMask * wMask);
  const bAdj = (blacks / 100.0) * 0.25 * Math.pow(bMask, 1.8);

  r = clamp(r * (1.0 + hAdj) + (1.0 - r) * sAdj + wAdj + bAdj);
  g = clamp(g * (1.0 + hAdj) + (1.0 - g) * sAdj + wAdj + bAdj);
  b = clamp(b * (1.0 + hAdj) + (1.0 - b) * sAdj + wAdj + bAdj);

  // 6. Tone Curve (Dreamcore Matte Curve)
  r = evaluateToneCurve(r);
  g = evaluateToneCurve(g);
  b = evaluateToneCurve(b);

  // 7. HSL Adjustments + Vibrance (+46) & Saturation (+49)
  let [h, s, l] = rgbToHsl(r, g, b);
  const weights = getHueWeights(h);

  const HSL_SETTINGS = {
    red:     { hue: 0.0,   sat: 0.0,   lum: 0.0 },
    orange:  { hue: 13.0,  sat: -13.0, lum: 0.0 },
    yellow:  { hue: -6.0,  sat: 6.0,   lum: -9.0 },
    green:   { hue: 0.0,   sat: 26.0,  lum: 0.0 },
    aqua:    { hue: 10.0,  sat: 14.0,  lum: 0.0 },
    blue:    { hue: 0.0,   sat: -13.0, lum: 11.0 },
    purple:  { hue: 0.0,   sat: 15.0,  lum: 0.0 },
    magenta: { hue: 0.0,   sat: 0.0,   lum: 0.0 }
  };

  let deltaH = 0;
  let deltaS = 0;
  let deltaL = 0;
  for (const [name, w] of Object.entries(weights)) {
    if (w > 0 && HSL_SETTINGS[name]) {
      deltaH += w * HSL_SETTINGS[name].hue;
      deltaS += w * HSL_SETTINGS[name].sat;
      deltaL += w * HSL_SETTINGS[name].lum;
    }
  }

  // Hue shift
  h = (h + (deltaH / 360.0) * 30.0 + 360.0) % 360.0;

  // Saturation with Global Saturation (+49) & Vibrance (+46)
  const saturationGlobal = 49.0;
  const vibrance = 46.0;
  let satMult = 1.0 + (saturationGlobal / 100.0) * 0.70 + (deltaS / 100.0) * 0.50;
  if (vibrance !== 0) {
    satMult += (vibrance / 100.0) * 0.50 * (1.0 - s);
  }
  s = clamp(s * Math.max(0.0, satMult));

  // Luminance shift
  l = clamp(l * (1.0 + (deltaL / 100.0) * 0.40));

  [r, g, b] = hslToRgb(h, s, l);
  return [clamp(r), clamp(g), clamp(b)];
}

function main() {
  const lutSize = 33;
  const outputPath = '/root/Dreamcore.cube';

  let lines = [];
  lines.push('# Created for Nothing Camera preset - Dreamcore');
  lines.push('TITLE "Dreamcore"');
  lines.push(`LUT_3D_SIZE ${lutSize}`);
  lines.push('DOMAIN_MIN 0.0 0.0 0.0');
  lines.push('DOMAIN_MAX 1.0 1.0 1.0');
  lines.push('');

  for (let bIdx = 0; bIdx < lutSize; bIdx++) {
    const bVal = bIdx / (lutSize - 1);
    for (let gIdx = 0; gIdx < lutSize; gIdx++) {
      const gVal = gIdx / (lutSize - 1);
      for (let rIdx = 0; rIdx < lutSize; rIdx++) {
        const rVal = rIdx / (lutSize - 1);
        const [outR, outG, outB] = processColor(rVal, gVal, bVal);
        lines.push(`${outR.toFixed(6)} ${outG.toFixed(6)} ${outB.toFixed(6)}`);
      }
    }
  }

  fs.writeFileSync(outputPath, lines.join('\n') + '\n', 'utf8');
  console.log(`Generated ${outputPath} successfully with ${lutSize}x${lutSize}x${lutSize} points.`);
}

main();

File: generate_dreamcore.py
#!/usr/bin/env python3
"""
Dreamcore LUT Generator for Nothing Camera
Constructs a high-precision 33x33x33 .CUBE 3D LUT matching exact Lightroom parameters.
"""

import numpy as np
import colorsys

def build_tone_curve_lut():
    # 6 control points modeling the Dreamcore lifted-black matte tone curve
    # Input -> Output points (0.0 to 1.0)
    # 1. Lifted blacks (matte shadow)
    # 2. Lower shadows lifted
    # 3. Soft midtone transition
    # 4. Controlled upper midtones
    # 5. Smooth highlight rolloff
    # 6. Slightly muted peak whites
    xp = np.array([0.00, 0.18, 0.38, 0.62, 0.82, 1.00], dtype=np.float32)
    fp = np.array([0.08, 0.22, 0.42, 0.63, 0.83, 0.97], dtype=np.float32)
    
    # 1024-step high-res 1D lookup table using linear/smooth monotonic interpolation
    x_eval = np.linspace(0.0, 1.0, 1024, dtype=np.float32)
    curve_1d = np.interp(x_eval, xp, fp)
    return curve_1d

def apply_tone_curve(rgb, curve_1d):
    idx = np.clip(rgb * 1023.0, 0, 1023).astype(np.int32)
    return curve_1d[idx]

def apply_white_balance(rgb, temp=-8.0, tint=5.0):
    # Temperature: negative = cooler/cyan-blue, positive = warmer/amber
    # Tint: negative = green, positive = magenta
    # Scale factors calibrated to standard Lightroom Kelvin/Tint adjustments
    r_scale = 1.0 + (temp / 100.0) * 0.12 + (tint / 100.0) * 0.08
    g_scale = 1.0 - (tint / 100.0) * 0.08
    b_scale = 1.0 - (temp / 100.0) * 0.12 - (tint / 100.0) * 0.04
    
    rgb = rgb * np.array([r_scale, g_scale, b_scale], dtype=np.float32)
    return np.clip(rgb, 0.0, 1.0)

def apply_light_parameters(rgb, exposure=0.60, contrast=-30.0, highlights=-54.0, shadows=34.0, whites=13.0, blacks=78.0, dehaze=-22.0):
    # 1. Exposure in EV stops
    rgb = rgb * (2.0 ** exposure)
    
    # 2. Dehaze approximation (-22 lowers midtone contrast and lifts shadows/blacks creating a dreamy haze)
    if dehaze != 0:
        haze_factor = (dehaze / 100.0) * 0.15
        rgb = rgb - haze_factor * (1.0 - rgb)
        rgb = np.clip(rgb, 0.0, 1.0)

    # 3. Basic Contrast (-30: compress toward mid-gray 0.45)
    c = (contrast / 100.0) * 0.40
    rgb = (rgb - 0.45) * (1.0 + c) + 0.45
    rgb = np.clip(rgb, 0.0, 1.0)
    
    # Calculate luminance for selective adjustments
    luma = 0.2126 * rgb[..., 0] + 0.7152 * rgb[..., 1] + 0.0722 * rgb[..., 2]
    
    # 4. Highlights (-54: pull back upper range)
    h_mask = np.clip((luma - 0.45) / 0.55, 0.0, 1.0)[..., np.newaxis]
    rgb = rgb + (highlights / 100.0) * 0.35 * (rgb * h_mask)
    
    # 5. Shadows (+34: lift lower-mid tones)
    s_mask = np.clip((0.55 - luma) / 0.55, 0.0, 1.0)[..., np.newaxis]
    rgb = rgb + (shadows / 100.0) * 0.30 * ((1.0 - rgb) * s_mask)
    
    # 6. Whites (+13: expand extreme highlights)
    w_mask = np.clip(luma / 1.0, 0.0, 1.0)[..., np.newaxis]
    rgb = rgb + (whites / 100.0) * 0.15 * (w_mask ** 2)
    
    # 7. Blacks (+78: significantly lift deep shadows / floor)
    b_mask = np.clip(1.0 - luma, 0.0, 1.0)[..., np.newaxis]
    rgb = rgb + (blacks / 100.0) * 0.25 * (b_mask ** 1.8)
    
    return np.clip(rgb, 0.0, 1.0)

def get_hue_weights(hue_deg):
    h = hue_deg % 360.0
    centers = {
        'red': 0.0,
        'orange': 30.0,
        'yellow': 60.0,
        'green': 120.0,
        'aqua': 180.0,
        'blue': 240.0,
        'purple': 285.0,
        'magenta': 325.0
    }
    weights = {}
    for name, center in centers.items():
        diff = abs(h - center)
        if diff > 180.0:
            diff = 360.0 - diff
        # Smooth cosine/triangular falloff
        weight = max(0.0, 1.0 - (diff / 35.0))
        weights[name] = weight
        
    total = sum(weights.values())
    if total > 0:
        for k in weights:
            weights[k] /= total
    return weights

def apply_color_mix(rgb, hsl_map, vibrance=46.0, saturation=49.0):
    flat_rgb = rgb.reshape(-1, 3)
    out_rgb = np.zeros_like(flat_rgb)
    
    sat_scale = 1.0 + (saturation / 100.0) * 0.70
    
    for i in range(len(flat_rgb)):
        r, g, b = flat_rgb[i]
        h, l, s = colorsys.rgb_to_hls(r, g, b)
        h_deg = h * 360.0
        
        weights = get_hue_weights(h_deg)
        
        delta_h = sum(weights[c] * hsl_map.get(c, {}).get('hue', 0.0) for c in weights)
        delta_s = sum(weights[c] * hsl_map.get(c, {}).get('sat', 0.0) for c in weights)
        delta_l = sum(weights[c] * hsl_map.get(c, {}).get('lum', 0.0) for c in weights)
        
        # 1. Hue shift
        h = (h + (delta_h / 360.0) * 0.30) % 1.0
        
        # 2. Saturation shift + Vibrance
        s_mult = sat_scale + (delta_s / 100.0) * 0.50
        if vibrance != 0:
            # Vibrance boosts muted colors more strongly
            vib_boost = (vibrance / 100.0) * 0.50 * (1.0 - s)
            s_mult += vib_boost
            
        s = np.clip(s * max(0.0, s_mult), 0.0, 1.0)
        
        # 3. Luminance shift
        l = np.clip(l * (1.0 + (delta_l / 100.0) * 0.40), 0.0, 1.0)
        
        new_r, new_g, new_b = colorsys.hls_to_rgb(h, l, s)
        out_rgb[i] = [new_r, new_g, new_b]
        
    return out_rgb.reshape(rgb.shape)

def generate_dreamcore_lut(output_path="/root/Dreamcore.cube", lut_size=33):
    # Create 33x33x33 base RGB lattice
    grid = np.linspace(0.0, 1.0, lut_size, dtype=np.float32)
    r_grid, g_grid, b_grid = np.meshgrid(grid, grid, grid, indexing='ij')
    rgb_cube = np.stack([r_grid, g_grid, b_grid], axis=-1)
    
    # Exact HSL data provided
    HSL_DATA = {
        'red':     {'hue': 0.0,   'sat': 0.0,   'lum': 0.0},
        'orange':  {'hue': 13.0,  'sat': -13.0, 'lum': 0.0},
        'yellow':  {'hue': -6.0,  'sat': 6.0,   'lum': -9.0},
        'green':   {'hue': 0.0,   'sat': 26.0,  'lum': 0.0},
        'aqua':    {'hue': 10.0,  'sat': 14.0,  'lum': 0.0},
        'blue':    {'hue': 0.0,   'sat': -13.0, 'lum': 11.0},
        'purple':  {'hue': 0.0,   'sat': 15.0,  'lum': 0.0},
        'magenta': {'hue': 0.0,   'sat': 0.0,   'lum': 0.0}
    }
    
    # 1. White Balance (-8 Temp, +5 Tint)
    cube = apply_white_balance(rgb_cube, temp=-8.0, tint=5.0)
    
    # 2. Light & Tonal Parameters
    cube = apply_light_parameters(
        cube,
        exposure=0.60,
        contrast=-30.0,
        highlights=-54.0,
        shadows=34.0,
        whites=13.0,
        blacks=78.0,
        dehaze=-22.0
    )
    
    # 3. Tone Curve (Dreamcore lifted black / matte shape)
    curve_1d = build_tone_curve_lut()
    cube = apply_tone_curve(cube, curve_1d)
    
    # 4. HSL Mix & Vibrance (+46) / Saturation (+49)
    cube = apply_color_mix(cube, HSL_DATA, vibrance=46.0, saturation=49.0)
    
    # 5. Write standard .CUBE file
    with open(output_path, 'w') as f:
        f.write(f'# Created for Nothing Camera preset - Dreamcore\n')
        f.write(f'TITLE "Dreamcore"\n')
        f.write(f'LUT_3D_SIZE {lut_size}\n')
        f.write('DOMAIN_MIN 0.0 0.0 0.0\n')
        f.write('DOMAIN_MAX 1.0 1.0 1.0\n\n')
        
        # .CUBE iteration standard: R fastest, G middle, B slowest
        for b in range(lut_size):
            for g in range(lut_size):
                for r in range(lut_size):
                    val = cube[r, g, b]
                    f.write(f"{val[0]:.6f} {val[1]:.6f} {val[2]:.6f}\n")
                    
    print(f"Generated {output_path} successfully ({lut_size}x{lut_size}x{lut_size} nodes).")

if __name__ == "__main__":
    generate_dreamcore_lut()
